File: server/webapp/WEB-INF/rails.new/app/presenters/api_v2/config/tasks/base_task_representer.rb
##########################################################################
# Copyright 2016 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV2
  module Config
    module Tasks
      class BaseTaskRepresenter < ApiV2::BaseRepresenter
        alias_method :task, :represented
        collection :run_if, embedded: false, exec_context: :decorator, expect_hash:  true
        property :on_cancel_config,
                 expect_hash:  true,
                 as:           :on_cancel,
                 exec_context: :decorator,
                 decorator:    OnCancelRepresenter,
                 class:        com.thoughtworks.go.config.OnCancelConfig


        def run_if
          task.getConditions().map { |condition| condition.to_s }
        end

        def run_if=(value)
          run_if_conditions= RunIfConfigs.new
          value.each { |condition|
            run_if_conditions.add(RunIfConfig.new(condition))
          }
          task.setConditions(run_if_conditions)
        end

        def on_cancel_config
          task.getOnCancelConfig() if task.hasCancelTask()
        end

        def on_cancel_config=(value)
          @represented.setOnCancelConfig(value)
        end
      end
    end
  end
end

File: api/api-shared-v11/src/main/java/com/thoughtworks/go/apiv11/admin/shared/representers/stages/tasks/BaseTaskRepresenter.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.apiv11.admin.shared.representers.stages.tasks;

import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.config.AbstractTask;
import com.thoughtworks.go.config.OnCancelConfig;
import com.thoughtworks.go.config.RunIfConfig;
import com.thoughtworks.go.domain.RunIfConfigs;

import java.util.stream.Collectors;

public class BaseTaskRepresenter {

    public static void toJSON(OutputWriter jsonWriter, AbstractTask task) {
        jsonWriter.addChildList("run_if", task.getConditions().stream().map(RunIfConfig::toString).collect(Collectors.toList()));
        if (task.hasCancelTask()) {
            jsonWriter.addChild("on_cancel", attributeWriter -> OnCancelRepresenter.toJSON(attributeWriter, task.getOnCancelConfig()));
        }
    }

    public static AbstractTask fromJSON(JsonReader jsonReader, AbstractTask task) {
        RunIfConfigs runIfConfigs = new RunIfConfigs();
        jsonReader.readArrayIfPresent("run_if", configs -> {
            configs.forEach(runIfConfig -> {
                runIfConfigs.add(new RunIfConfig(runIfConfig.getAsString()));
            });
        });
        task.setConditions(runIfConfigs);
        jsonReader.optJsonObject("on_cancel").ifPresent(onCancelReader -> {
            OnCancelConfig onCancelConfig = OnCancelRepresenter.fromJSON(jsonReader.readJsonObject("on_cancel"));
            task.setOnCancelConfig(onCancelConfig);
        });
        return task;
    }
}
